xpti/include/xpti/xpti_trace_framework.hpp

```diff
@@ -585,33 +597,162 @@ class tracepoint_t {
   // Constructor that makes calls to xpti API layer to register strings and
   // create the Universal ID that is stored in the TLS entry for lookup
   tracepoint_t(xpti::payload_t *p) : m_payload(nullptr), m_top(false) {
-    if (p) {
-      // We expect the payload input has been populated with the information
-      // available at that time
-      uint64_t uid = xptiGetUniversalId();
+    // If tracing is not enabled, don't do anything
+    if (!xptiTraceEnabled())
+      return;
+
+    init();
+    // We expect the payload input has been populated with the information
+    // available at that time; before we use this payload, we need to check if a
+    // tracepoint has been set at a higher scope.
+    uint64_t uid = xptiGetUniversalId();
+    if (uid != xpti::invalid_uid) {
+      // We already have a parent SW layer that has a tracepoint defined. This
+      // should be associated with a trace event and a payload
+      m_trace_event =
+          const_cast<xpti::trace_event_data_t *>(xptiFindEvent(uid));
+      // If the trace event is valid, extract the payload
+      if (m_trace_event) {
+        m_payload = m_trace_event->reserved.payload;
+      } else {
+        // Trace event is unavailable, so let is create one with the payload
+        // associated with the UID;
+        m_payload = xptiQueryPayloadByUID(uid);
+        m_trace_event = xptiMakeEvent(
+            m_default_name, const_cast<xpti::payload_t *>(m_payload),
+            m_default_event_type, m_default_activity_type, &m_instID);
+      }
+    } else if (p) {
```

Comment:
Likely a stupid question, but why is this in the same PR as the introduction of TLS? I would have expected we need mostly new code, not extensive modifications to the existing one.

Reply:
Initial TLS implementation committed over a year ago always assumed that the TLS contained a 64-bt universal ID that was to be used by the rest of the SW layers in the scope. However, the ability to stash code location was added to allow the notification to be decoupled from instrumenting the SYCL entry points. This allows to continue to close feature gaps without breaking ABI and enable code location information when ABI breakage is allowed.